src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -825,24 +882,45 @@ mongoc_collection_estimated_document_count (
    }
 
    reply_ptr = reply ? reply : &reply_local;
-   bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
-
-   ret = _mongoc_client_command_with_opts (coll->client,
-                                           coll->db,
-                                           &cmd,
-                                           MONGOC_CMD_READ,
-                                           opts,
-                                           MONGOC_QUERY_NONE,
-                                           read_prefs,
-                                           coll->read_prefs,
-                                           coll->read_concern,
-                                           coll->write_concern,
-                                           reply_ptr,
-                                           error);
-
-   if (ret) {
-      if (bson_iter_init_find (&iter, reply_ptr, "n")) {
-         count = bson_iter_as_int64 (&iter);
+   if (server_stream->sd->max_wire_version < WIRE_VERSION_4_9) {
+      /* On < 4.9, use actual count command for estimatedDocumentCount */
+      bson_init (&cmd);
+      bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
+      ret = _mongoc_client_command_with_opts (coll->client,
+                                             coll->db,
+                                             &cmd,
+                                             MONGOC_CMD_READ,
+                                             opts,
+                                             MONGOC_QUERY_NONE,
+                                             read_prefs,
+                                             coll->read_prefs,
+                                             coll->read_concern,
+                                             coll->write_concern,
+                                             reply_ptr,
+                                             error);
+      if (ret) {
+         if (bson_iter_init_find (&iter, reply_ptr, "n")) {
+            count = bson_iter_as_int64 (&iter);
+         }
+      }
+   } else {
+      /* On >= 4.9, use aggregate with collStats for estimatedDocumentCount */
+      _make_aggregate_for_edc (coll, &cmd);
+      ret = mongoc_collection_read_command_with_opts (
+         coll, &cmd, read_prefs, opts, reply_ptr, error);
+
+      if (error && error->code == MONGOC_ERROR_COLLECTION_DOES_NOT_EXIST) {
+         /* Collection does not exist. From spec: return 0 but no err:
+          * https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#estimateddocumentcount
+          */
+         error = NULL;
+         count = 0;
+         GOTO (done);
+      }
+      if (ret && bson_iter_init(&iter, reply_ptr)) {
+         if (bson_iter_find_descendant (&iter, "cursor.firstBatch.0.n", &iter)) {
+            count = bson_iter_as_int64 (&iter);
+         }
```

Comment:
If the command was successful but we don't find the data we're looking for, we're returning a count of `-1` without exposing any additional information. It may be sensible to set `error` to something that indicates that we didn't receive the information we were looking for.

Reply:
An interesting point! The estimated document count function did not return an error on a successful but malformed server response before this added functionality, so I'm not sure what a meaningful error would be. Any thoughts @bazile-clyde ?

Reply:
Since this is an extremely narrow use-case (only thing I can imagine is the server not returning an initial batch in the cursor), I'm fine leaving this as is and revisiting as part of CDRIVER-3933 or another ticket.